### ISSUE TITLE
Dependency Management: Poetry

### ISSUE THREAD

**Issue by mjones-oddball** [department-of-veterans-affairs/notification-api#1665]
## User Story - Business Need
We want to use Poetry as our dependency manager, to make use of several of the features it has such as the lockfile, dependency resolution, parallel downloads, and simplification of CI.

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

As VA Notify
I want to simplify the dependency management of VA Notify
So that we can quickly and effectively keep our system up to date with the latest changes

### Additional Info and Resources

- [Poetry docs](https://python-poetry.org/docs/)

## Engineering Checklist

- [x] Determine how Poetry would replace the requirements-app.text, requirements.txt, requirements_for_local.txt, requirements_for_test.txt and the requirements files for the lambdas, which are just used as documentation.
- [x] Determine what changes would be necessary to build the actions/containers to use poetry instead of pip
- [x] Implement modifications determined above
- [x] Modify Github actions as necessary
- [x] Additional ticket(s) are created if more work is needed to implement than can be done in the allotted points
- [ ] Update or close relevant SPIKE: #1202

## Acceptance Criteria

- [x] Dependency tracking
> GIVEN Poetry is set up for package management in notification-api
WHEN I implement changes and build
THEN I can track the state of my dependencies

- [x] No requirements files remain
- [x] Updated readme
- [x] Updated API dependency template with the new process

**Comment by k-macmillan** [comment]
Separated local run from test run. Tests are now running. Currently stuck with database deadlocks with the following failures. 
```
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_billing_triggers_tasks_for_days[None-expected_kwargs0] - AssertionError: assert 0 == 4
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_billing_triggers_tasks_for_days[2019-07-21-expected_kwargs1] - AssertionError: assert 0 == 4
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_notification_status_triggers_tasks_for_days[None-expected_kwargs0] - AssertionError: assert 0 == 4
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_notification_status_triggers_tasks_for_days[2019-07-21-expected_kwargs1] - AssertionError: assert 0 == 4
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_notification_status_triggers_tasks_for_days_including_csv_generation_when_feature_flag_on[None-expected_kwargs0] - AssertionError: assert 0 == 4
FAILED tests/app/celery/test_reporting_tasks.py::test_create_nightly_notification_status_triggers_tasks_for_days_including_csv_generation_when_feature_flag_on[2019-07-21-expected_kwargs1] - AssertionError: assert 0 == 4
FAILED tests/app/delivery/test_send_to_providers.py::test_notification_raises_error_if_message_contains_sin_pii_that_passes_luhn - app.exceptions.InvalidProviderException: Could not determine a provider using strategy LOAD_BALANCING.
FAILED tests/app/delivery/test_send_to_providers.py::test_client_to_use_should_return_template_provider - app.exceptions.InvalidProviderException: The provider 15344d95-1fea-42cc-a089-cd03d23b50f3 could not be found.
FAILED tests/app/delivery/test_send_to_providers.py::test_should_not_use_active_but_disabled_provider - AssertionError: Regex pattern did not match.
FAILED tests/app/celery/test_tasks.py::test_should_save_sms_if_restricted_service_and_valid_number - AssertionError: assert 'technical-failure' == 'created'
FAILED tests/app/notifications/rest/test_send_notification.py::test_should_send_email_to_anyone_with_test_key[0-True] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/notifications/test_process_notifications.py::test_persist_notification_persists_recipient_identifiers[PID-some pid-email] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/v2/notifications/test_post_notifications.py::test_post_sms_notification_returns_201[data1-reference_from_client] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/v2/notifications/test_post_notifications.py::test_post_sms_notification_uses_sms_sender_id_reply_to - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/v2/notifications/test_post_notifications.py::test_post_notification_with_scheduled_for[email-email_address-sample@email.com] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/performance_platform/test_total_sent_notifications.py::test_get_total_sent_notifications_yesterday_returns_expected_totals_dict - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/v2/notifications/test_post_notifications.py::test_should_not_persist_or_send_notification_if_simulated_recipient[simulate-delivered@notifications.va.gov-email] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/notifications/test_process_notifications.py::test_persist_sms_notification_stores_normalised_number[  6502532223-+16502532223] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/celery/test_tasks.py::test_should_save_sms_template_to_and_persist_with_job_id - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/v2/notifications/test_post_notifications.py::test_post_email_notification_with_invalid_reply_to_id_returns_400 - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/notifications/rest/test_send_notification.py::test_should_not_send_notification_if_restricted_and_not_a_service_user[email-not-someone-we-trust@email-address.com] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
FAILED tests/app/dao/notification_dao/test_notification_dao.py::test_dao_get_notifications_by_to_field_escapes[%-2] - sqlalchemy.exc.OperationalError: (psycopg2.errors.DeadlockDetected) deadlock detected
ERROR tests/app/notifications/rest/test_send_notification.py::test_should_not_send_notification_for_archived_template - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_should_not_send_notification_for_archived_template> (7 sub-exceptions)
ERROR tests/app/dao/notification_dao/test_notification_dao.py::test_dao_get_notifications_by_to_field_escapes[%_-1] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_dao_get_notifications_by_to_field_escapes[%_-1]> (2 sub-exceptions)
ERROR tests/app/dao/notification_dao/test_notification_dao.py::test_dao_get_notifications_by_to_field_escapes[%-2] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_dao_get_notifications_by_to_field_escapes[%-2]> (10 sub-exceptions)
ERROR tests/app/v2/notifications/test_post_notifications.py::test_post_notification_with_scheduled_for[email-email_address-sample@email.com] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_post_notification_with_scheduled_for[email-email_address-sample@email.com]> (3 sub-exceptions)
ERROR tests/app/v2/notifications/test_post_notifications.py::test_should_not_persist_or_send_notification_if_simulated_recipient[simulate-delivered@notifications.va.gov-email] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_should_not_persist_or_send_notification_if_simulated_recipient[simulate-delivered@notifications.va.gov-...
ERROR tests/app/notifications/rest/test_send_notification.py::test_should_not_send_notification_if_restricted_and_not_a_service_user[email-not-someone-we-trust@email-address.com] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_should_not_send_notification_if_restricted_and_not_a_service_user[email-not-someone-we-trust@email-addr...
ERROR tests/app/notifications/test_process_notifications.py::test_persist_notification_persists_recipient_identifiers[PID-some pid-email] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_persist_notification_persists_recipient_identifiers[PID-some pid-email]> (3 sub-exceptions)
ERROR tests/app/v2/notifications/test_post_notifications.py::test_post_sms_notification_uses_sms_sender_id_reply_to - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_post_sms_notification_uses_sms_sender_id_reply_to> (3 sub-exceptions)
ERROR tests/app/notifications/test_process_notifications.py::test_persist_sms_notification_stores_normalised_number[  6502532223-+16502532223] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_persist_sms_notification_stores_normalised_number[  6502532223-+16502532223]> (3 sub-exceptions)
ERROR tests/app/notifications/rest/test_send_notification.py::test_should_send_email_to_anyone_with_test_key[0-True] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_should_send_email_to_anyone_with_test_key[0-True]> (3 sub-exceptions)
ERROR tests/app/v2/notifications/test_post_notifications.py::test_post_email_notification_with_invalid_reply_to_id_returns_400 - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_post_email_notification_with_invalid_reply_to_id_returns_400> (3 sub-exceptions)
ERROR tests/app/v2/notifications/test_post_notifications.py::test_post_sms_notification_returns_201[data1-reference_from_client] - exceptiongroup.ExceptionGroup: errors while tearing down <Function test_post_sms_notification_returns_201[data1-reference_from_client]> (3 sub-exceptions)
```

**Comment by k-macmillan** [comment]
Stopped using a multi-stage build in case it was missing a .so or something. Didn't help. 

Seems to have been related to the way celery interacts with the reworked containers. Had to rewrite celery assertions.

Hit a wall with unit tests getting stuck in a deadlock. This only happens with multi-worker. Refactored a lot of tests, could not resolve it. Rolled xdist back a few versions and it fixed it. The final hook, `pytest_sessionfinish`, appears to have been executed at the first worker being done, while the other workers were still running. Didn't trust that because the API currently uses xdist 3.5. Rebuilt everything with 3.5 again and it is working. Something must have hit a state it could not recover from.

Running just the parallel tests with an army of workers:
<img width="997" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN1doQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1a5ddf55c53e76c51a71712f4b303a121bc3e3c2/image.png" alt="image.png" />

Container running with the same command we'll give it when deployed passes all tests.
<img width="437" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN21oQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--802dc1c00f7cf5303586506add2a5a4b20f9af64/image.png" alt="image.png" />


Next-up:
- Meet with @mchlwellman to see if his his one-file work can be implemented
- Adjust non-local Dockerfiles so it can be deployed. Need to address the Github Actions too.

**Comment by k-macmillan** [comment]
It started failing again. Found templates that were not being torn down correctly, fixing those appear to have fixed it. Built new Dockerfile for deployed environment, but it's currently working locally too. Deleted other Dockerfiles (local and test). docker compose can be used to call the same Dockerfile so we should have a more consistent experience with testing locally vs on github.

Currently passing all tests locally. Need to get with @ldraney  because I cannot test these in the cloud. The workflow is not capable of dispatch, so I can't manually exercise it.

**Comment by mjones-oddball** [comment]
Updating from a 5 to an 8 after discussion about off track.

**Comment by k-macmillan** [comment]
Tests are executing with Poetry during PRs now. The tests are failing pretty bad (every `F` is a failure), so I will have to tweak the pyproject.toml and compare to what we have locked down in the rep currently. 

<img width="472" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK08wQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--36bacf738b6174cea12c6dc99e7583c891238de9/image.png" alt="image.png" />

**Comment by k-macmillan** [comment]
Decided to take a more narrow approach and focused just on `tests/app/celery`. Properly tore down everything that looked wrong and marked any tests that were failing for parallel worker reasons as `serial`.  Still got some ft_billing deadlocks locally, and [passes all unit tests in the PR](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8180219780/job/22367802737?pr=1667) (with far more deadlocks).

Putting deadlocks on hold for now. All of the changes made were good changes that handle things better but it may be possible that one of the updated packages is causing this. 

Moving to deploy & regression testing first thing tomorrow.

**Comment by k-macmillan** [comment]
Image size in ECR is less than half that of current images:
![image](https://github.com/department-of-veterans-affairs/notification-api/assets/16893311/6e859073-410f-4256-a8c6-d8481d77a82b)

**Comment by k-macmillan** [comment]
Made revisions to allow for migrations to pass. Now having permission errors related to save_certificate.sh. Cleaned out all old requirements references (aside from the README.md).

**Comment by k-macmillan** [comment]
TLDR: Over the hill!

Resolved permission errors, but it took a while to navigate what part of the process required what permissions. Added comments to the Dockerfile. Fails to deploy with a lot of `eventlet`/`greenlet` errors so I removed `eventlet` as a test and it deployed and passed most of regression (notification was sent but the regression suite didn't get that for some reason). 

Tried putting eventlet back in along with some other changes, didn't work. Downgraded eventlet and dnspython to what we currently use in the repo. Still had issues. Downgraded ddtrace to 1.18.2, fixed it.

All but 2 regression tests passed. The two that "failed" show `delivered` status when looking up their notification IDs. I ran the QA suite a second time to verify it was not just due to too short of a window between when the deploy finished and when QA did the lookup. I can send myself email/sms through postman. Not sure what is up with the QA repo but will need to talk with QA tomorrow. 

Remaining steps:
- I regularly updated with `master` but today had multiple PRs that would have been disruptive, so I need to pull recent changes in. One of them is likely to have conflicts
- Ensure all unit tests are passing in the PRs after the changes
- Probably deploy to perf as well to make sure it works across different environments before possibly breaking `master`
- Update readme

**Comment by k-macmillan** [comment]
Attempting to deal with api_key table deadlocks. Probably going to mark all associated tests serial and add a comment to them, then move on.

**Comment by k-macmillan** [comment]
Commented out entire files, was still having issues. Walked the tests one subfolder at a time, fixing issues as I went. Took most of the day but it was [worth the effort](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8240179516). No longer seeing deadlocks after extensive modifications to tests.

<img width="1587" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMXpQQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--772a21d281534b07391d2e1fa1fe19786ef0d143/image.png" alt="image.png" />


Passing all [unit tests](https://github.com/department-of-veterans-affairs/notification-api/actions/runs/8240177172) and [regression tests](https://github.com/department-of-veterans-affairs/notification-api-qa/actions/runs/8240247124).

TODO:
- Thoroughly review all changes to ensure necessity/validity
- Update README.md
- Ensure CodeQL and dependency scanning is executing properly

**Comment by k-macmillan** [comment]
Deadlocks appeared to have gone away but they resurfaced after changes to a README file, indicating that we were just "lucky" to not have the deadlocks before. I have pulled this out of "Ready for Review" status. 

In an attempt to isolate the problem I created a new branch and reverted all changes in the tests folder and still have deadlocks. 

Met with @kalbfled for some time yesterday and we eliminated tests/app/conftest.py cases where we called session.delete(ORM object), and changed them all to delete().where() queries. Did not resolve the deadlocks.

**Comment by k-macmillan** [comment]
A quirk with the way workers execute tasks is causing the cleanup function to run early. Some workers will hit a long-running function but still have many tasks allocated. The remaining tasks exceed the sleep timer originally designed to handle the issue of workers still performing tests. This seems heavily related to post methods using the `client` fixture, or any internal "post" to the API. These tests can sometimes take dozens of seconds to run.

Moved the truncate behind a much longer sleep timer and it stopped having deadlocks. Added a variable (TRUNCATE_ARTIFACTS) to `.local.env` to only truncate if set to True and set the sleep timer to 5 by default. 

While working on a ticket the truncate functionality is a very valuable tool. Unfortunately, it is currently unsafe to leave enabled at all times as it once was, so devs will have to turn it on if they are working on individual unit tests.

From the pytest [documentation](https://pytest-xdist.readthedocs.io/en/latest/how-to.html#making-session-scoped-fixtures-execute-only-once):
> pytest-xdist is designed so that each worker process will perform its own collection and execute a subset of all tests.

And it appears after the first worker has finished all of their tests it kicks off the `pytest_sessionfinish` hook. 

&nbsp;

This is passing all Github-ran unit tests and the Regression Suite. I addressed a few PR comments and have stated it is ready for re-review.

**Comment by mjones-oddball** [comment]
@k-macmillan Per our convo today I added AC "Updated API dependency template with the new process"

**Comment by k-macmillan** [comment]
Updated based on PR comments and requested re-review.

**Comment by k-macmillan** [comment]
Created a [pytest.mark.serial cleaning ticket](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1690) and a [dependency template update ticket](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/notification-api/1692).

PR has been merged.
